### PR TITLE
MPI: Add `run_exports`

### PR DESCRIPTION
**Pull request by ax3l**
Ensure pinning of compile-time MPI to downstream.

cc @minrk for help :pray: 

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by ax3l** [comment]
@conda-forge-admin, please rerender

**Review by leofang — CHANGES_REQUESTED** [review]
This is incorrect because it's the job of mpi packages (in this case, `impi-rt`) to define run_exports, not their consumers (otherwise everyone would have to define run_exports). Check how it's done in `openmpi`/`mpich`.

I noticed that `msmpi` does not define run_exports either, but it's another bug to fix.

**Comment by leofang** [comment]
> Isn't this issue the kind of situations where repodata patches are preferred?

No, unfortunately run_exports can't be patched (yet). We need to do this properly, and mark all impacted packages (impi & mpi4py) as broken.

**Comment by leofang** [comment]
For some reason this morning I missed that `impi-devel` already added `impi_rt` to run_exports, so once we fix `impi_rt` to add the mutex (the repodata patch is [already done](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/579), but we still need to [fix the recipe](https://github.com/conda-forge/intel_repack-feedstock/pull/57)), we should be good.

**Comment by ax3l** [comment]
Thank you all for this. Please let me know if this PR needs to be updated (or is not needed at all?) after these upstream PRs are merged.

**Comment by ax3l** [comment]
> This is incorrect because it's the job of mpi packages (in this case, impi-rt) to define run_exports, not their consumers (otherwise everyone would have to define run_exports).

@leofang thanks for the guidance. But isn't defining these run exports exactly what the knowledge base tells me to do?
https://conda-forge.org/docs/maintainer/knowledge_base.html#building-mpi-variants

Or does this not apply, because mpi4py does not have a preferred MPI provider (and def. no `nompi` variant)?

**Comment by minrk** [comment]
_using_ mpi4py doesn't imply requiring a specific mpi implementation. _linking_ against that implementation does, though. So if you do link a specific mpi implementation, that implementation needs to be in your `host` requirements, which should take care of the run_exports.

As in the knowledge base, passing along the run_exports is appropriate for libraries that may be linked to, where the link will break if the mpi implementation changes (e.g. petsc).

repodata patches for run_exports aren't hugely valuable, since run_exports necessarily only affect _future_ builds of downstream packages linking _outdated_ versions of the package that was patched, which doesn't happen that often.

**Comment by leofang** [comment]
As @minrk noted, adding `run_exports` is not needed for mpi4py (and is incorrect, because nothing links to mpi4py, and it'd only make the solver's life harder). Let's close this PR (but thanks @ax3l for trying to help 🙂).

**Pull request closed without merging**
